bin/cheritest/cheritest_printf.c

```diff
@@ -0,0 +1,220 @@
+	if ((cheri_getperm(p) & CHERI_PERM_EXECUTE) != 0) {
+		if (*permsp != 'x')
+			cheritest_failure_errx("Missing 'x' permission for %s",
+			    descr);
+		permsp++;
+	}
+	if ((cheri_getperm(p) & CHERI_PERM_LOAD) != 0) {
+		if (*permsp != 'R')
+			cheritest_failure_errx("Missing 'R' permission for %s",
+			    descr);
+		permsp++;
+	}
+	if ((cheri_getperm(p) & CHERI_PERM_STORE) != 0) {
+		if (*permsp != 'W')
+			cheritest_failure_errx("Missing 'W' permission for %s",
+			    descr);
+		permsp++;
+	}
+	if (*permsp != '\0')
+		cheritest_failure_errx("Extra permissions '%s' for %s", permsp,
+		    descr);
+
+	if (base != cheri_getbase(p))
+		cheritest_failure_errx("Mismatched base for %s", descr);
+	if (top != cheri_getbase(p) + cheri_getlength(p))
+		cheritest_failure_errx("Mismatched top for %s", descr);
+
+	if (tokens == 4)
+		return;
+
+	if (cheri_gettag(p)) {
+		if (strstr(attr, "invalid") != NULL)
+			cheritest_failure_errx("Tagged cap marked invalid "
+			    "for %s", descr);
+	} else {
+		if (strstr(attr, "invalid") == NULL)
+			cheritest_failure_errx("Untagged cap not marked "
+			    "invalid for %s", descr);
+	}
+
+	switch (cheri_gettype(p)) {
+	case CHERI_OTYPE_UNSEALED:
+		if (strstr(attr, "sealed") != NULL)
+			cheritest_failure_errx("Unsealed cap marked as "
+			    "sealed for %s", descr);
+		if (strstr(attr, "sentry") != NULL)
+			cheritest_failure_errx("Unsealed cap marked as "
+			    "sentry for %s", descr);
+		break;
+	case CHERI_OTYPE_SENTRY:
+		if (strstr(attr, "sealed") != NULL)
+			cheritest_failure_errx("Sentry cap marked as "
+			    "sealed for %s", descr);
+		if (strstr(attr, "sentry") == NULL)
+			cheritest_failure_errx("Sentry cap not marked as "
+			    "sentry for %s", descr);
+		break;
+	default:
+		if (strstr(attr, "sealed") == NULL)
+			cheritest_failure_errx("Sealed cap not marked as "
+			    "sealed for %s", descr);
+		if (strstr(attr, "sentry") != NULL)
+			cheritest_failure_errx("Sealed cap marked as "
+			    "sentry for %s", descr);
+		break;
+	}
+}
+
+void
+test_printf_cap(const struct cheri_test *ctp __unused)
+{
+	char data[64];
+	void * __capability scalar = (void * __capability)(uintcap_t)4;
+	char * __capability datap = data;
+
+	snprintf(data, sizeof(data), "%#lp", CAP_ARG(scalar));
+	if (strcmp(data, "0x4") != 0)
+		cheritest_failure_errx("Wrong output for simple scalar");
+
+	snprintf(data, sizeof(data), "%#.4lp", CAP_ARG(scalar));
+	if (strcmp(data, "0x0004") != 0)
+		cheritest_failure_errx("Wrong output for simple scalar "
+		    "with precision");
+
+	snprintf(data, sizeof(data), "%#8lp", CAP_ARG(scalar));
+	if (strcmp(data, "     0x4") != 0)
+		cheritest_failure_errx("Wrong output for simple scalar "
+		    "with padding");
+
+	snprintf(data, sizeof(data), "%#-8lp", CAP_ARG(scalar));
+	if (strcmp(data, "0x4     ") != 0)
+		cheritest_failure_errx("Wrong output for simple scalar "
+		    "with left adjust padding");
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wformat"
```

Comment:
What's this for?

Reply:
clang warns about the '0' prefix which the code ignores.  I could just drop the test instead perhaps.  Testing the fact that we ignore flags we don't define as legal is perhaps a bit much.

Reply:
Yeah I'd just drop it then.

Reply:
Will do.